src/daft-core/src/array/ops/utf8.rs

```diff
@@ -1422,6 +1422,19 @@ impl Utf8Array {
             .with_validity(self_arrow.validity().cloned());
         Ok(Self::from((self.name(), Box::new(arrow_result))))
     }
+
+    pub fn encode<Encoder>(&self, encoder: Encoder) -> DaftResult<BinaryArray>
+    where
+        Encoder: Fn(&str) -> DaftResult<Cow<'_, [u8]>>,
+    {
+        let arr = self.as_arrow();
+        let res = arr
+            .iter()
+            .map(|val| val.map_or(Ok(None), |s| encoder(s).map(Some)))
+            .collect::<DaftResult<arrow2::array::BinaryArray<i64>>>()?
+            .with_validity(arr.validity().cloned());
+        Ok(BinaryArray::from((self.name(), Box::new(res))))
+    }
```

Comment:
the utf8 arrays are backed by a `Buffer<u8>` _(`[u8]`)_, and are already utf8 validated. No need to convert it back to a `&str` when we can just operate on the raw bytes. `encoder` should instead just take in `&[u8]` instead of `&str`